### DISCUSSION TITLE
promtool and openmetrics pitfalls

### DISCUSSION THREAD

**Discussion by lucab** [prometheus/prometheus#9661] [Q&A] [answered]
For context, it looks like there is a different definition of timestamps between Prometheus and OpenMetrics exposition formats:
 * Prometheus [mandates](https://prometheus.io/docs/instrumenting/exposition_formats/#comments-help-text-and-type-information) Unix epoch **milliseconds** as a signed **integer**
 * OpenMetrics [specified](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#numbers) Unix epoch **seconds** as a **float** (with optional nanoseconds part)

This is a small enough detail that it gets easily missed by most people.

To make things even worst, there are some `promtool` verbs that are not aware of Prometheus own format and thus force users into this pitfall through a lossy roundtrip via OpenMetrics (e.g. `promtool tsdb create-blocks-from openmetrics`).
I've seen at least two occurrences of this in the wild just this year:
 * https://www.mail-archive.com/prometheus-users@googlegroups.com/msg06570.html
 * https://github.com/tikv/rust-prometheus/issues/425

Is there a design reason behind this `promtool` verb only handling OpenMetrics formats?
Would a `promtool` RFE for native support of Prometheus own format be in scope for that tool (e.g. `promtool tsdb create-blocks-from prometheus`)?

**Comment by beorn7** [comment, marked as answer]
The goal is essentially to converge onto OpenMetrics as the newer and more general format. `promtool tsdb create-blocks-from` is a new feature, so it was implemented using OpenMetrics from the beginning. Prometheus itself prefers to scrape in OpenMetrics format if the scrape target offers it.

What you call the native Prometheus format is simply the "old" format. And it had milliseconds because back then it was meant to be Prometheus specific, and Prometheus just offers millisecond precision, so it made sense. OpenMetrics is meant to be a universally useful format, so it doesn't make any assumptions about the timestamp precision of the target TSDB.

To remove confusion, we should deprecate more clearly the uses of the old Prometheus text format rather than adding more uses of it.

**Comment by lucab** [comment]
Thanks for the feedback! Then from what I can take away from your reply, the answers to my doubts are:
 * yes, it's a design decision to let `promtool tsdb create-blocks-from` only handle OpenMetrics format (and not the old Prometheus format)
 * no, such RFE would go against the intended future direction, so it's very unlikely to happen

**Reply by beorn7** [reply]
I believe that's a correct description of the current state.
